README.md
# Badges
[![Code style check](https://github.com/blooo-io/LedgerHQ-app-plugin-squid/actions/workflows/lint-workflow.yml/badge.svg?branch=main)](https://github.com/blooo-io/LedgerHQ-app-plugin-squid/actions/workflows/lint-workflow.yml)
[![Compilation & tests](https://github.com/blooo-io/LedgerHQ-app-plugin-squid/actions/workflows/ci-workflow.yml/badge.svg?branch=main)](https://github.com/blooo-io/LedgerHQ-app-plugin-squid/actions/workflows/ci-workflow.yml)

# Ledger squid Plugin

This is a plugin for the Ethereum application which helps parsing and displaying relevant information when signing a squid transaction.

## Prerequisite

Clone the plugin in a new folder.

```shell
git clone https://github.com/blooo-io/LedgerHQ-app-plugin-squid.git
```

Then in the same folder clone two more repositories, which is the plugin-tools and app-ethereum.

```shell
git clone https://github.com/LedgerHQ/plugin-tools.git                          #plugin-tools
git clone --recurse-submodules https://github.com/LedgerHQ/app-ethereum.git     #app-ethereum
```
## Documentation

Need more information about the interface, the architecture, or general stuff about ethereum plugins? You can find more about them in the [ethereum-app documentation](https://github.com/LedgerHQ/app-ethereum/blob/master/doc/ethapp_plugins.asc).

## Smart Contracts

The smart contracts covered by this plugin are:

| Network   | Smart Contract |
| ---       | ---            |
| Ethereum  | `0xce16F69375520ab01377ce7B88f5BA8C48F8D666`|
| Polygon   | `0xce16F69375520ab01377ce7B88f5BA8C48F8D666`|
| BSC       | `0xce16F69375520ab01377ce7B88f5BA8C48F8D666`|


On these smart contracts, the functions covered by this plugin are:

|    Function   | Selector  | Displayed Parameters |
| ---           | ---       | --- |
|callBridgeCall | 0x8ca3bf68| <table>  <tbody>  <tr> <td><code>address token</code></td></tr> <tr><td><code>uint256 amount</code></td></tr> <tr><td><code>string destinationChain</code></td></tr> <tr><td><code>string bridgedTokenSymbol</code></td></tr> </tbody> </table> |
|bridgeCall     | 0x3c659741| <table>  <tbody>  <tr><td><code>string destinationChain</code></td></tr> <tr><td><code>string bridgedTokenSymbol</code></td></tr> <tr><td><code>uint256 amount</code></td></tr> </tbody> </table>|
|callBridge     | 0xf35af1f8| <table>  <tbody>  <tr> <td><code>address token</code></td></tr> <tr><td><code>uint256 amount</code></td></tr> <tr><td><code>string destinationChain</code></td></tr> <tr><td><code>string bridgedTokenSymbol</code></td></tr> </tbody> </table> |


## Build

Go to the plugin-tools folder and run the "./start" script.
```shell
cd plugin-tools  # go to plugin folder
./start.sh       # run the script start.sh
```
The script will build a docker image and attach a console.
When the docker image is running go to the "LedgerHQ-app-plugin-squid" folder and build the ".elf" files.
```shell
cd LedgerHQ-app-plugin-squid/tests    # go to the tests folder in app-plugin-squid
./build_local_test_elfs.sh            # run the script build_local_test_elfs.sh
```

## Tests

To test the plugin go to the tests folder from the "app-plugin-squid" and run the script "test"
```shell
cd LedgerHQ-app-plugin-squid/tests    # go to the tests folder in LedgerHQ-app-plugin-squid
yarn test                             # run the script test
```
## Continuous Integration


The flow processed in [GitHub Actions](https://github.com/features/actions) is the following:

- Code formatting with [clang-format](http://clang.llvm.org/docs/ClangFormat.html)
- Compilation of the application for Ledger Nano S, Nano S+ and Nano X in [ledger-app-builder](https://github.com/LedgerHQ/ledger-app-builder)
- Execution of the provided end to end tests

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2022-12-16

### Added

- Inital version of the brand new squid plugin for Ethereum application.


src/contract.c
#include "squid_plugin.h"

// Need more information about the interface for plugins? Please read the README.md!

// You can check SquidRouterProxy's methods here :
// https://etherscan.io/address/0xce16f69375520ab01377ce7b88f5ba8c48f8d666#code
//

// callBridgeCall : 0x8ca3bf68
static const uint8_t SQUID_CALL_BRIDGE_CALL_SELECTOR[SELECTOR_SIZE] = {0x8c, 0xa3, 0xbf, 0x68};

// bridgeCall : 0x3c659741
static const uint8_t SQUID_BRIDGE_CALL_SELECTOR[SELECTOR_SIZE] = {0x3c, 0x65, 0x97, 0x41};

// callBridge : 0xf35af1f8
static const uint8_t SQUID_CALL_BRIDGE_SELECTOR[SELECTOR_SIZE] = {0xf3, 0x5a, 0xf1, 0xf8};

// Array of all the different Squid selectors.
const uint8_t *const SQUID_SELECTORS[NUM_SQUID_SELECTORS] = {SQUID_CALL_BRIDGE_CALL_SELECTOR,
                                                             SQUID_BRIDGE_CALL_SELECTOR,
                                                             SQUID_CALL_BRIDGE_SELECTOR};

const uint8_t NULL_ETH_ADDRESS[ADDRESS_LENGTH] = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
                                                  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
                                                  0x00, 0x00, 0x00, 0x00, 0x00, 0x00};

// Array containing the chains supported by squid
const struct chainIdToChainName_t SQUID_SUPPORTED_CHAINS[NUM_SUPPORTED_CHAINS] = {
    {"Ethereum", "Ethereum"},
    {"Arbitrum", "Arbitrum"},
    {"Avalanche", "Avalanche C-Chain"},
    {"Moonbeam", "Moonbeam Mainnet"},
    {"Polygon", "Polygon"},
    {"binance", "Binance"},
    {"cosmoshub", "Cosmos Hub"},
    {"crescent", "Crescent"},
    {"injective", "Injective"},
    {"juno", "Juno"},
    {"kujira", "Kujira"},
    {"osmosis", "Osmosis"},
    {"secret", "Secret Network"},
    {"terra-2", "Terra-2"},
    {"Agoric", "Agoric"},
    {"ASSETMANTLE", "AssetMantle"},
    {"Axelarnet", "Axelar"},
    {"COMDEX", "Comdex"},
    {"EVMOS", "Evmos"},
    {"fetch", "Fetch"},
    {"KI", "Ki"},
    {"REGEN", "Regen"},
    {"UMEE", "Umee"},
};

// Array containing the most used tokens supported by squid
// each represented with their symbol and decimals
const struct tokenSymbolToDecimals_t SQUID_SUPPORTED_TOKENS[NUM_SUPPORTED_TOKENS] = {
    {"ETH", 18},
    {"WETH", 18},
    {"DAI", 18},
    {"USDT", 6},
    {"USDC", 6},
    {"WBTC", 8},
    {"MATIC", 18},
    {"BNB", 18},
    {"BUSD", 18},
    {"wAXL", 6},
    {"axlUSDC", 6},
    {"axlDAI", 18},
    {"axlUSDT", 6},
    {"WMATIC", 18},
    {"WBNB", 18},
    {"WGLMR", 18},
    {"WFTM", 18},
    {"WAVAX", 18},
};